Split AccountsGrooves into mutable and immutable Grooves

kprotty opened the pull request:
Splits out the immutable and mutable portions of the Account struct into their own Grooves to decrease write amplification overhead. `AccountImmutableGroove` contains all the non-balance fields with an IdTree while `AccountsMutableGroove` contains the balance fields updated frequently by `create_transfers`. Changes to Groove are made to conditionally opt-out of IdTree usage.

Closes #255 

## Pre-merge checklist

(TODO) Performance:

* ~~[ ] Compare `zig benchmark` on linux before and after this pr.~~
    ``` sh
    On my system, the standard deviation in req/s is too large to accurately tell if it introduces a regression or not :/
    ```

Commit:
lsm: make Groove "id" optional, defaulting to timestamps

There will be Grooves for Objects that dont use the id as
a primary key indirection for the timestamp. These grooves
should therefore not have any intermediary IdTree overhead.

Commit:
sm: split the AccountsGroove into {Im}Mutable grooves

Most of the account fields don't change after creation.
Only the debits/credits counters do on create_transfer.
Keeping the immutable portions of Accounts in their own
Grooves should help decrease write amplification overhead.

sentientwaffle commented on src/lsm/groove.zig:
```diff
@@ -310,8 +309,8 @@ pub fn GrooveType(
         groove_options.ignored.len -
-        // The id/timestamp field is implicitly ignored since it's the primary key for ObjectTree:
-        2 +
+        // The id/timestamp fields is implicitly ignored since it's the primary key for ObjectTree:
```
Grammar.

sentientwaffle commented on src/lsm/groove.zig:
```diff
@@ -381,24 +380,26 @@ pub fn GrooveType(
 
-        const PrefetchIDs = std.AutoHashMapUnmanaged(u128, void);
+        const id_field = if (use_id) "id" else "timestamp";
+        const Id = @TypeOf(@field(@as(Object, undefined), id_field));
```
I suggest using a different name for these (`id_field` and `Id`) to avoid overloading "id".

Maybe `primary_field`/`Primary`?

Commit:
sm: reduce size of AccountImmutable & code changes

Commit:
sm: call prefetch on AccountsMutableGroove for create_accounts

Not calling prefetch() when put_no_clobber() is called in
`commit(.create_accounts)` led to the Groove's `prefetch_object`
map not being cleared on `prefetch_setup()`, hitting an assert.

Commit:
zig fmt

Commit:
sm: expose Account(Im)Mutable & fix tests

Exposing them allows other components like
forest_fuzz to interact with the grooves.

Commit:
lsm: fix forest_fuzz to work with Account(Im)Mutable

Commit:
sm: get timestamp from inserted account on rollback

The account passed into `create_account_rollback` doesn't
have a proper `timestamp` set which was being used to remove
the AccountMutable. Instead, fetch the timestamp from the
existing inserted AccountImmutable.

also zig fmt on forest_fuzz

Commit:
lsm: always deinit prefetch_ids in errdefer

sentientwaffle commented on src/state_machine.zig:
```diff
@@ -31,11 +31,83 @@ pub fn StateMachineType(comptime Storage: type, comptime constants_: struct {
+            code: u16,
+            flags: AccountFlags,
+            padding: [16]u8,
+
```
Add `comptime` assertions explicitly verifying that `@sizeOf()` is the expected value, and `@bitSizeOf()` has zero implicit padding. (See other `extern struct`s for examples).

sentientwaffle commented on src/lsm/forest_fuzz.zig:
```diff
@@ -194,23 +194,66 @@ const Environment = struct {
+        const groove_mut = &env.forest.grooves.accounts_mutable;
+
+        const GrooveImmut = @TypeOf(groove_immut.*);
+        const GrooveMut = @TypeOf(groove_mut.*);
```
I think for the groove types we should spell out "immutable"/"mutable", though I'm ok with abbreviating it in the variable names since it is repeated so often.

sentientwaffle replied:
Missed from previous review?

sentientwaffle commented on src/lsm/forest_fuzz.zig:
```diff
@@ -194,23 +194,66 @@ const Environment = struct {
+            prefetch_context_mut: GrooveMut.PrefetchContext = undefined,
+            prefetch_context_immut: GrooveImmut.PrefetchContext = undefined,
+
+            fn prefetch_start(getter: *@This()) void {
```
`s/@This()/Getter` (more explicit, and makes it easier to navigate/search code)

kprotty replied:
Getter can't refer to itself without `@This()` as its defined in a function rather than a namespace. Should it be moved just outside `prefetch_account`?

sentientwaffle replied:
Ahh no its fine I just didn't realize that was how zig's scoping rules worked.

sentientwaffle commented on src/lsm/groove.zig:
```diff
@@ -526,7 +527,7 @@ pub fn GrooveType(
             // be in the manifest but not yet on disk).
             const snapshot_max = groove.objects.lookup_snapshot_max;
-            assert(snapshot_max == groove.ids.lookup_snapshot_max);
+            if (use_id) assert(snapshot_max == groove.ids.lookup_snapshot_max);
```
Maybe `assert(!use_id or snapshot_max == groove.ids.lookup_snapshot_max);`?

sentientwaffle commented on src/lsm/groove.zig:
```diff
@@ -546,27 +547,31 @@ pub fn GrooveType(
-                if (id_tree_value.tombstone()) {
+        pub fn prefetch_enqueue(groove: *Groove, id: Id) void {
+            if (use_id) {
+                if (groove.ids.lookup_from_memory(groove.prefetch_snapshot.?, id)) |id_tree_value| {
```
This might read better as

```
if (!use_id) {
    groove.prefetch_ids.putAssumeCapacity(id, {});
    return;
}
// fall through to original logic.
```

sentientwaffle commented on src/lsm/groove.zig:
```diff
@@ -157,8 +155,9 @@ pub fn GrooveType(
 
-    assert(@hasField(Object, "id"));
-    assert(std.meta.fieldInfo(Object, .id).field_type == u128);
+    const use_id = @hasField(Object, "id");
```
nit: instead of `use_id`, would `has_id` be more accurate?

sentientwaffle commented on src/lsm/groove.zig:
```diff
@@ -762,7 +774,7 @@ pub fn GrooveType(
         fn insert(groove: *Groove, object: *const Object) void {
             groove.objects.put(object);
-            groove.ids.put(&IdTreeValue{ .id = object.id, .timestamp = object.timestamp });
+            if (use_id) groove.ids.put(&IdTreeValue{ .id = @field(object, id_field), .timestamp = object.timestamp });
```
This can be `object.id` rather than using `id_field` since it is guarded by `use_id`.

sentientwaffle commented on src/lsm/groove.zig:
```diff
@@ -262,7 +261,7 @@ pub fn GrooveType(
     };
 
-    const IdTree = blk: {
+    const IdTree = if (!use_id) void else blk: {
```
What if, rather than using `if (use_id)` conditions everywhere we need an `IdTree`, there was instead a dummy `IdTree` implementation for id-less trees. Most functions would be noops, but `lookup_from_memory` would always "succeed" (since it takes a timestamp and returns a timestamp).

(Idk if the above ends up being cleaner, but worth a try.)

kprotty replied:
I'll try that out

kprotty replied:
Considered it, but I needed to mock up even the asynchronous bits like open()/compact()/checkpoint() and it didn't seem worth it.

sentientwaffle commented on src/state_machine.zig:
```diff
@@ -31,11 +31,83 @@ pub fn StateMachineType(comptime Storage: type, comptime constants_: struct {
+            ledger: u32,
+            code: u16,
+            flags: AccountFlags,
+            padding: [16]u8,
```
Should we add another 8 bytes here so that the mutable/immutable paddings add up to `Account.reserved`?

kprotty replied:
Originally, Immutable had a larger `reserved` to account for the padding but keeping them smaller made more sense to reduce space overhead.

sentientwaffle replied:
But if in the future we want to use all of `reserved`, won't we need to put it somewhere?

kprotty replied:
We would then put it into Immutable/Mutable as applicable, updating the padding to account (no pun intended) for the new fields. I assume `reserved` serves as both padding and type-erased implementation storage for Transfer/Account, where Immutable/Mutable would *be* the implementation with fields more specific than `reserved`.

sentientwaffle replied:
Ok, but unless my math is wrong we are 8 bytes short. Adding those changes the disk format, so it is something we want to accommodate in advance.

kprotty replied:
Had some more time to think about this. Adding 8 more bytes to `AccountImmutable.padding` would no longer make the struct a power of two size (which is comptime asserted down the line by `SetAssociativeCache`). Also, I think the other 8 bytes are accounted for with `AccountMutable` having a copy of `timestamp` (vs. a single Account struct having only one `timestamp`)

sentientwaffle replied:
> Had some more time to think about this. Adding 8 more bytes to AccountImmutable.padding would no longer make the struct a power of two size (which is comptime asserted down the line by SetAssociativeCache).

:+1: Ok, that's fine then.  But just fyi:

> Also, I think the other 8 bytes are accounted for with AccountMutable having a copy of timestamp (vs. a single Account struct having only one timestamp)

Since those 8 bytes are "duplicated", they don't actually solve the problem. We still wind up with 8 bytes of `reserved` that we don't have room to store. But its fine for now.

sentientwaffle commented on src/state_machine.zig:
```diff
@@ -263,16 +338,29 @@ pub fn StateMachineType(comptime Storage: type, comptime constants_: struct {
+        }
+
+        fn prefetch_create_accounts_immutable_callback(completion: *AccountsImmutableGroove.PrefetchContext) void {
+            const self = @fieldParentPtr(StateMachine, "prefetch_accounts_immutable_context", completion);
```
wrap long lines

sentientwaffle commented on src/state_machine.zig:
```diff
@@ -263,16 +338,29 @@ pub fn StateMachineType(comptime Storage: type, comptime constants_: struct {
-        fn prefetch_create_accounts_callback(completion: *AccountsGroove.PrefetchContext) void {
-            const self = @fieldParentPtr(StateMachine, "prefetch_accounts_context", completion);
+        fn prefetch_create_accounts_mutable_callback(completion: *AccountsMutableGroove.PrefetchContext) void {
+            const self = @fieldParentPtr(StateMachine, "prefetch_accounts_mutable_context", completion);
```
wrap long lines

sentientwaffle commented on src/state_machine.zig:
```diff
@@ -334,17 +448,32 @@ pub fn StateMachineType(comptime Storage: type, comptime constants_: struct {
+
+            const ids = mem.bytesAsSlice(Event(.lookup_accounts), self.prefetch_input.?);
+            for (ids) |id| {
+                const immut = self.forest.grooves.accounts_immutable.get(id).?;
```
wrt `.?`, how is this guaranteed to exist?

kprotty replied:
Good point, Should be a `orelse continue`.

sentientwaffle replied:
Can you also figure out why the state machine's unit tests didn't catch this, and ideally add one that would?

kprotty replied:
As mentioned externally, the state machine's unit tests don't call `prefetch()` given they seem to only test accounting logic. Was able to hit the assert I believe with VOPR seed `13017736572448279926` before the vopr failed to no longer build the simulator binary.

sentientwaffle commented on src/state_machine.zig:
```diff
@@ -577,8 +706,9 @@ pub fn StateMachineType(comptime Storage: type, comptime constants_: struct {
+                    const mut = self.forest.grooves.accounts_mutable.get(immut.timestamp).?;
+                    results[results_count] = into_account(immut, mut);
                     results_count += 1;
                 }
```
```
} else {
    assert(self.forest.grooves.accounts_mutable.get(immut.timestamp) == null);
}
```

kprotty replied:
`immut` isn't available by then as it's the result of the `if` lookup

sentientwaffle replied:
Ah right, nevermind

sentientwaffle commented on src/state_machine.zig:
```diff
@@ -624,23 +754,30 @@ pub fn StateMachineType(comptime Storage: type, comptime constants_: struct {
             if (@bitCast(u16, a.flags) != @bitCast(u16, e.flags)) return .exists_with_different_flags;
             if (a.user_data != e.user_data) return .exists_with_different_user_data;
-            assert(zeroed_48_bytes(a.reserved) and zeroed_48_bytes(e.reserved));
+            assert(zeroed_48_bytes(a.reserved));
```
Assert `AccountImmutable` and `AccountMutable` padding is `0`.

sentientwaffle commented on src/state_machine.zig:
```diff
@@ -971,20 +1121,19 @@ pub fn StateMachineType(comptime Storage: type, comptime constants_: struct {
                             batch_accounts_max,
-                            // ×2 because creating a transfer will update 2 accounts.
-                            2 * batch_transfers_max,
+                            batch_transfers_max,
```
nit: does the `math.max(..., ...)` fit on one line now?

sentientwaffle commented on src/state_machine.zig:
```diff
@@ -995,6 +1144,26 @@ pub fn StateMachineType(comptime Storage: type, comptime constants_: struct {
+                        .cache_entries_max = options.cache_entries_accounts,
+                        .commit_entries_max = math.max(
+                            batch_accounts_max,
+                            // ×2 because creating a transfer will update 2 AccountsMutable.
```
```suggestion
                            // ×2 because creating a transfer will update 2 AccountMutable's.
```

since we are referring to the item, not the groove.

Commit:
lsm/sm: add Account(Im)mutable asserts and padding asserts

Commit:
lsm: rename Groove's use_id to has_id

Commit:
LSM: rename id_field/Id to primary_field/PrimaryKey in Groove

For instances like `AccountsMutableGroove` in `StateMachine`,
the primary key is the `timestamp` field instead of the `id` field
so it made less sense to call it `id` on both occasions in source.

Commit:
sm: wrap long lines

Commit:
sm: skip AccountImmutables not found in prefetch

After prefetching `AccountImmutable`s, the corresponding
`AccountMutable`s are also prefetched. The loop to do this
incorrectly assumed all requested `AccountImmutable`s for
`.lookup_accounts` were found/valid which is not always true.

sentientwaffle commented on src/lsm/groove.zig:
```diff
@@ -170,7 +169,7 @@ pub fn GrooveType(
         // By default, we ignore the "timestamp" field since it's a special identifier.
         // Since the "timestamp" is ignored by default, it shouldn't be provided in groove_options.ignored.
-        comptime var ignored = mem.eql(u8, field.name, "timestamp") or mem.eql(u8, field.name, "id");
+        comptime var ignored = mem.eql(u8, field.name, "timestamp") or (has_id and mem.eql(u8, field.name, "id"));
```
I think `has_id or` is redundant here — if `mem.eql(u8, field.name, "id")` is true, then `has_id` is (by definition) true as well.

sentientwaffle commented on src/lsm/groove.zig:
```diff
@@ -526,7 +527,7 @@ pub fn GrooveType(
             // be in the manifest but not yet on disk).
             const snapshot_max = groove.objects.lookup_snapshot_max;
-            assert(snapshot_max == groove.ids.lookup_snapshot_max);
+            assert(!has_id or snapshot_max == groove.ids.lookup_snapshot_max);
```
Readability nit: could you swap the conditions (to read `assert(snapshot_max == groove.ids.lookup_snapshot_max or !has_id);`) so that `snapshot_max` "aligns" with the preceding line?

kprotty replied:
the `has_id` before acts as a comptime short-circuit to avoid accessing `groove.ids` if it doesn't exist. Would something like `if (has_id) { \n assert()` be better?

sentientwaffle replied:
Ah, forgot about that. You can leave it as-is.

sentientwaffle reviewed (commented):
Just a few minor changes left, but other than that this looks great!

Commit:
lsm: groove/id naming fixups

sentientwaffle commented:
bors r+

sentientwaffle commented:
bors retry

sentientwaffle commented:
bors retry

sentientwaffle commented:
bors r-

Seems like this is an actual bug.

batiati commented:
The java test [testCreateInvalidTransfer](https://github.com/tigerbeetledb/tigerbeetle/blob/main/src/clients/java/src/test/java/com/tigerbeetle/IntegrationTest.java#L409-L436)  hangs due to a server crash:

```
Formatting replica ...
info(io): creating "0_0.tigerbeetle.examples"...
info(io): allocating 1.25994873046875GiB...
info(main): 0: formatted: cluster=0
Starting tigerbeetle ...
info(io): opening "0_0.tigerbeetle.examples"...
info(main): 0: Allocated 1643655168 bytes in 12 regions during replica init
info(main): 0: cluster=0: listening on 127.0.0.1:3001
info(message_bus): connection from client 127099763929920966269809350402148742505
thread 939115 panic: attempt to use null value
/home/batiati/tigerbeetle/src/state_machine.zig:438:100: 0x3e1bbd in .vsr.state_machine.StateMachineType(.vsr.storage.Storage,(struct .vsr.state_machine.struct:25:70 constant)).prefetch_create_transfers_callback_accounts_immutable (tigerbeetle)
                    const dr_immut = self.forest.grooves.accounts_immutable.get(t.debit_account_id).?;
                                                                                                   ^
/home/batiati/tigerbeetle/src/lsm/groove.zig:642:33: 0x4307cf in .vsr.lsm.groove.PrefetchContext.finish (tigerbeetle)
                context.callback(context);
                                ^
/home/batiati/tigerbeetle/src/lsm/groove.zig:632:62: 0x402889 in .vsr.lsm.groove.PrefetchContext.worker_finished (tigerbeetle)
                if (context.workers_busy == 0) context.finish();
                                                             ^
/home/batiati/tigerbeetle/src/lsm/groove.zig:627:40: 0x3e10ff in .vsr.lsm.groove.PrefetchContext.start_workers (tigerbeetle)
                context.worker_finished();
                                       ^
/home/batiati/tigerbeetle/src/lsm/groove.zig:592:34: 0x3c31c5 in .vsr.lsm.groove.GrooveType(.vsr.storage.Storage,.vsr.state_machine.AccountImmutable,(struct .vsr.state_machine.struct:106:14 constant)).prefetch (tigerbeetle)
            context.start_workers();
                                 ^
/home/batiati/tigerbeetle/src/state_machine.zig:419:60: 0x3c37bc in .vsr.state_machine.StateMachineType(.vsr.storage.Storage,(struct .vsr.state_machine.struct:25:70 constant)).prefetch_create_transfers_callback_transfers (tigerbeetle)
            self.forest.grooves.accounts_immutable.prefetch(
                                                           ^
/home/batiati/tigerbeetle/src/lsm/groove.zig:642:33: 0x4312df in .vsr.lsm.groove.PrefetchContext.finish (tigerbeetle)
                context.callback(context);
                                ^
/home/batiati/tigerbeetle/src/lsm/groove.zig:632:62: 0x4035a9 in .vsr.lsm.groove.PrefetchContext.worker_finished (tigerbeetle)
                if (context.workers_busy == 0) context.finish();
                                                             ^
/home/batiati/tigerbeetle/src/lsm/groove.zig:627:40: 0x3e186f in .vsr.lsm.groove.PrefetchContext.start_workers (tigerbeetle)
                context.worker_finished();
                                       ^
/home/batiati/tigerbeetle/src/lsm/groove.zig:592:34: 0x3c3655 in .vsr.lsm.groove.GrooveType(.vsr.storage.Storage,.vsr.tigerbeetle.Transfer,(struct .vsr.state_machine.struct:122:14 constant)).prefetch (tigerbeetle)
            context.start_workers();
                                 ^
/home/batiati/tigerbeetle/src/state_machine.zig:397:51: 0x3a4f0a in .vsr.state_machine.StateMachineType(.vsr.storage.Storage,(struct .vsr.state_machine.struct:25:70 constant)).prefetch_create_transfers (tigerbeetle)
            self.forest.grooves.transfers.prefetch(
                                                  ^
/home/batiati/tigerbeetle/src/state_machine.zig:318:51: 0x389493 in .vsr.state_machine.StateMachineType(.vsr.storage.Storage,(struct .vsr.state_machine.struct:25:70 constant)).prefetch (tigerbeetle)
                    self.prefetch_create_transfers(mem.bytesAsSlice(Transfer, input));
                                                  ^
/home/batiati/tigerbeetle/src/vsr/replica.zig:2225:40: 0x373487 in .vsr.vsr.replica.ReplicaType(.vsr.state_machine.StateMachineType(.vsr.storage.Storage,(struct .vsr.state_machine.struct:25:70 constant)),.vsr.message_bus.MessageBusType(.vsr.ProcessType.replica),.vsr.storage.Storage,.vsr.time.Time).commit_op_prefetch (tigerbeetle)
            self.state_machine.prefetch(
                                       ^
/home/batiati/tigerbeetle/src/vsr/replica.zig:2551:40: 0x380d1b in .vsr.vsr.replica.ReplicaType(.vsr.state_machine.StateMachineType(.vsr.storage.Storage,(struct .vsr.state_machine.struct:25:70 constant)),.vsr.message_bus.MessageBusType(.vsr.ProcessType.replica),.vsr.storage.Storage,.vsr.time.Time).commit_pipeline_next (tigerbeetle)
                self.commit_op_prefetch(prepare.message, commit_pipeline_callback);
                                       ^
/home/batiati/tigerbeetle/src/vsr/replica.zig:2526:38: 0x36b5a0 in .vsr.vsr.replica.ReplicaType(.vsr.state_machine.StateMachineType(.vsr.storage.Storage,(struct .vsr.state_machine.struct:25:70 constant)),.vsr.message_bus.MessageBusType(.vsr.ProcessType.replica),.vsr.storage.Storage,.vsr.time.Time).commit_pipeline (tigerbeetle)
            self.commit_pipeline_next();
                                     ^
/home/batiati/tigerbeetle/src/vsr/replica.zig:1021:33: 0x34f8f5 in .vsr.vsr.replica.ReplicaType(.vsr.state_machine.StateMachineType(.vsr.storage.Storage,(struct .vsr.state_machine.struct:25:70 constant)),.vsr.message_bus.MessageBusType(.vsr.ProcessType.replica),.vsr.storage.Storage,.vsr.time.Time).on_prepare_ok (tigerbeetle)
            self.commit_pipeline();
                                ^
/home/batiati/tigerbeetle/src/vsr/replica.zig:731:50: 0x2da272 in .vsr.vsr.replica.ReplicaType(.vsr.state_machine.StateMachineType(.vsr.storage.Storage,(struct .vsr.state_machine.struct:25:70 constant)),.vsr.message_bus.MessageBusType(.vsr.ProcessType.replica),.vsr.storage.Storage,.vsr.time.Time).on_message (tigerbeetle)
                .prepare_ok => self.on_prepare_ok(message),
                                                 ^
/home/batiati/tigerbeetle/src/vsr/replica.zig:2979:32: 0x36ba20 in .vsr.vsr.replica.ReplicaType(.vsr.state_machine.StateMachineType(.vsr.storage.Storage,(struct .vsr.state_machine.struct:25:70 constant)),.vsr.message_bus.MessageBusType(.vsr.ProcessType.replica),.vsr.storage.Storage,.vsr.time.Time).flush_loopback_queue (tigerbeetle)
                self.on_message(message);
                               ^
/home/batiati/tigerbeetle/src/vsr/replica.zig:5720:44: 0x38ddc3 in .vsr.vsr.replica.ReplicaType(.vsr.state_machine.StateMachineType(.vsr.storage.Storage,(struct .vsr.state_machine.struct:25:70 constant)),.vsr.message_bus.MessageBusType(.vsr.ProcessType.replica),.vsr.storage.Storage,.vsr.time.Time).write_prepare_callback (tigerbeetle)
            defer self.flush_loopback_queue();
                                           ^
/home/batiati/tigerbeetle/src/vsr/journal.zig:1817:27: 0x3c9c7b in .vsr.vsr.journal.JournalType(.vsr.vsr.replica.ReplicaType(.vsr.state_machine.StateMachineType(.vsr.storage.Storage,(struct .vsr.state_machine.struct:25:70 constant)),.vsr.message_bus.MessageBusType(.vsr.ProcessType.replica),.vsr.storage.Storage,.vsr.time.Time),.vsr.storage.Storage).write_prepare_release (tigerbeetle)
            write_callback(replica, wrote, write_trigger);
                          ^
/home/batiati/tigerbeetle/src/vsr/journal.zig:1785:42: 0x3e7b72 in .vsr.vsr.journal.JournalType(.vsr.vsr.replica.ReplicaType(.vsr.state_machine.StateMachineType(.vsr.storage.Storage,(struct .vsr.state_machine.struct:25:70 constant)),.vsr.message_bus.MessageBusType(.vsr.ProcessType.replica),.vsr.storage.Storage,.vsr.time.Time),.vsr.storage.Storage).write_prepare_on_write_header (tigerbeetle)
            journal.write_prepare_release(write, message);
                                         ^
/home/batiati/tigerbeetle/src/vsr/journal.zig:1931:27: 0x3e79ba in .vsr.vsr.journal.JournalType(.vsr.vsr.replica.ReplicaType(.vsr.state_machine.StateMachineType(.vsr.storage.Storage,(struct .vsr.state_machine.struct:25:70 constant)),.vsr.message_bus.MessageBusType(.vsr.ProcessType.replica),.vsr.storage.Storage,.vsr.time.Time),.vsr.storage.Storage).write_sectors_on_write (tigerbeetle)
            range.callback(write);
                          ^
/home/batiati/tigerbeetle/src/storage.zig:369:27: 0x3670dc in .vsr.storage.Storage.on_write (tigerbeetle)
            write.callback(write);
                          ^
/home/batiati/tigerbeetle/src/io/linux.zig:868:29: 0x367310 in .vsr.io.linux.struct:866:25.wrapper (tigerbeetle)
                    callback(
                            ^
/home/batiati/tigerbeetle/src/io/linux.zig:460:40: 0x33bca8 in .vsr.io.linux.Completion.complete (tigerbeetle)
                    completion.callback(completion.context, completion, &result);
                                       ^
/home/batiati/tigerbeetle/src/io/linux.zig:110:64: 0x339838 in .vsr.io.linux.IO.flush (tigerbeetle)
            while (copy.pop()) |completion| completion.complete();
                                                               ^
/home/batiati/tigerbeetle/src/io/linux.zig:90:27: 0x2cf672 in .vsr.io.linux.IO.run_for_ns (tigerbeetle)
            try self.flush(1, &timeouts, &etime);
                          ^
/home/batiati/tigerbeetle/src/tigerbeetle/main.zig:189:38: 0x2b4ed9 in Command.start (tigerbeetle)
            try command.io.run_for_ns(constants.tick_ms * std.time.ns_per_ms);
                                     ^
/home/batiati/tigerbeetle/src/tigerbeetle/main.zig:49:44: 0x2ac29d in main (tigerbeetle)
        .start => |*args| try Command.start(&arena, args),
                                           ^
/home/batiati/tigerbeetle/zig/lib/std/start.zig:561:37: 0x2a56da in std.start.callMain (tigerbeetle)
            const result = root.main() catch |err| {
                                    ^
/home/batiati/tigerbeetle/zig/lib/std/start.zig:495:12: 0x28627e in std.start.callMainWithArgs (tigerbeetle)
    return @call(.{ .modifier = .always_inline }, callMain, .{});
           ^
/home/batiati/tigerbeetle/zig/lib/std/start.zig:409:17: 0x285316 in std.start.posixCallMainAndExit (tigerbeetle)
    std.os.exit(@call(.{ .modifier = .always_inline }, callMainWithArgs, .{ argc, argv, envp }));
                ^
/home/batiati/tigerbeetle/zig/lib/std/start.zig:322:5: 0x285122 in std.start._start (tigerbeetle)
    @call(.{ .modifier = .never_inline }, posixCallMainAndExit, .{});
    ^
./scripts/run_tigerbeetle.sh: line 19: 939115 Aborted                 (core dumped) ../../../tigerbeetle start --addresses=3001 "$FILE"
```

kprotty committed:
sm: don't assume prefetched Account(Im)Mutables exist

kprotty committed:
sm: zig fmt

sentientwaffle commented:
bors d+

kprotty commented:
bors r+
